aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -515,6 +515,369 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
+// TODO: this implemtation is not efficient, it simply iteratates over
+// batch elements in a serial for-loop.
+template <typename func_t>
+void batch_apply(at::ArrayRef<Tensor> tensors, int64_t n_batch_dims, const func_t& f) {
+  // no-op if batch is empty
+  if (n_batch_dims <= 0) {
+    return;
+  }
+
+  std::vector<Tensor> tensors_view;
+  for (auto& tensor : tensors) {
+    std::vector<int64_t> tensor_view_sizes = {-1};
+    for (auto d = n_batch_dims; d < tensor.dim(); ++d) {
+      tensor_view_sizes.push_back(tensor.size(d));
+    }
+    auto tensor_view = tensor.view(tensor_view_sizes);
+    tensors_view.push_back(tensor_view);
+  }
+
+  for (int64_t input_idx = 0; input_idx < tensors_view[0].size(0);
+    ++input_idx) {
+    std::vector<Tensor> input_selected;
+    for (auto& tensor_view : tensors_view) {
+      auto curr_input = tensor_view.select(0, input_idx);
+      input_selected.push_back(curr_input);
+    }
+    f(input_selected);
+  }
+}
+
+Tensor matrix_power(const Tensor& matrices, const Tensor& powers) {
+  if (matrices.dim() > 2) {
+    auto res = at::empty(matrices.sizes(), matrices.options());
+
+    batch_apply(
+      {res, matrices, powers},
+      matrices.dim() - 2,
+      [](auto tensors) {
+        auto& out = tensors[0];
+        auto& in = tensors[1];
+        int64_t n = tensors[2].item().template to<int64_t>();
+        out.copy_(at::matrix_power(in, n));
+      }
+    );
+
+    return res;
+  }
+  else {
+    int64_t n = powers.item().template to<int64_t>();
+    return at::matrix_power(matrices, n);
+  }
+}
+
+Tensor operator_1_norm(const Tensor& tensor) {
+  return std::get<0>(tensor.abs().sum(-2).max(-1));
+}
+
+Tensor compute_T1(const Tensor& A) {
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  return I + A;
+}
+
+Tensor compute_T2(const Tensor& A) {
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  const auto A2 = at::matmul(A, A);
+  return I + A + A2 / fact[2];
+}
+
+Tensor compute_T4(const Tensor& A) {
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  const auto A2 = at::matmul(A, A);
+  return I + A + at::matmul(A2, I / fact[2] + A / fact[3] + A2 / fact[4]);
+}
+
+template <typename scalar_t>
+Tensor compute_T8(const Tensor& A) {
+  constexpr scalar_t sqrt_177 = 0.1330413469565007072504e+2;
+  constexpr scalar_t x3 = 2. / 3.;
+  constexpr scalar_t x1 = x3 * ((1. + sqrt_177) / 88.);
+  constexpr scalar_t x2 = x3 * ((1. + sqrt_177) / 352.);
+  constexpr scalar_t x4 = (-271. + 29. * sqrt_177) / (315. * x3);
+  constexpr scalar_t x5 = (-11. + 11. * sqrt_177) / (1260. * x3);
+  constexpr scalar_t x6 = (-99. + 11. * sqrt_177) / (5040. * x3);
+  constexpr scalar_t x7 = (89. - sqrt_177) / (5040. * x3);
+  constexpr scalar_t y2 = (857. - 58. * sqrt_177) / 630.;
+
+  const auto I = at::eye(A.size(-1), A.options()).expand_as(A);
+  const auto A2 = at::matmul(A, A);
+  const auto A4 = at::matmul(A2, x1 * A + x2 * A2);
```

Comment:
You could use `addcmul`. It performs addition and multiplication I think.

Reply:
Done!

Reply:
Need matrix product here, not pointwise, unfortunately.